### PR TITLE
Provide an option to allow annotate by gene query only in the indicator utils

### PR DESCRIPTION
**Pull request by zhx828**
This fixes https://github.com/oncokb/oncokb/issues/3688

### CODE REVIEW

File: core/src/main/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtils.java [hunk 1]
```diff
@@ -0,0 +1,708 @@
...
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.collections.map.HashedMap;
+import org.apache.commons.lang3.StringUtils;
+import org.mskcc.cbio.oncokb.apiModels.CancerTypeMatch;
+import org.mskcc.cbio.oncokb.apiModels.DrugMatch;
+import org.mskcc.cbio.oncokb.apiModels.LevelsOfEvidenceMatch;
+import org.mskcc.cbio.oncokb.genomenexus.GNVariantAnnotationType;
+import org.mskcc.cbio.oncokb.model.*;
+
+public class AnnotationSearchUtils {
```
Comment (zhx828) [inline core/src/main/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtils.java:21]: No need to review this file. All copied from the original AnnotationsApiController so I can test some logics.

File: core/src/test/java/org/mskcc/cbio/oncokb/util/IndicatorUtilsTest.java [hunk 2]
```diff
@@ -787,6 +787,48 @@ public void testProcessQuery() throws Exception {
...
+    @Test
+    public void testProcessQueryGeneOnlyQuery(){
+        // when alteration is specified in the query, the geneQueryOnly param doesn't make any difference
+        Query queryWithAlt = new Query(null, DEFAULT_REFERENCE_GENOME, null, "BRAF", "V123M", null, null, null, null, null, null, null);
+        IndicatorQueryResp indicatorQueryRespGeneOnly = IndicatorUtils.processQuery(queryWithAlt, null, true, null, true);
+        IndicatorQueryResp indicatorQueryResp = IndicatorUtils.processQuery(queryWithAlt, null, true, null, false);
+        pairComparison(indicatorQueryRespGeneOnly, indicatorQueryResp);
+
+        // Test when alteration is empty string, different geneOnlyQuery will return different result
+        Query query = new Query(null, DEFAULT_REFERENCE_GENOME, null, "BRAF", "", null, null, null, null, null, null, null);
```
[the remark this answers is not in the document]
Comment (zhx828) [reply core/src/test/java/org/mskcc/cbio/oncokb/util/IndicatorUtilsTest.java:799]: Certainly

File: core/src/main/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtils.java [hunk 3]
```diff
@@ -112,6 +112,13 @@ private static List<TypeaheadSearchResp> findMatchingCancerTypes(String query, B
         if (exactMatch) {
             matchedTumorTypes.add(ApplicationContextSingleton.getTumorTypeBo().getByName(query));
         } else {
+            for (Map.Entry<String, TumorType> currSubtype : CacheUtils.getCodedTumorTypeMap().entrySet()) {
+                String code = query.toUpperCase();
+                if (currSubtype.getKey().equals(code)) {
```
Comment (calvinlu3) [inline core/src/main/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtils.java:117]: The code should never have lowercase, so don't need to convert the key to uppercase here.

**Review by calvinlu3 — APPROVED** [review]
LGTM

File: core/src/main/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtils.java (hunk 3 above)
Comment (zhx828) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtils.java:117]: @calvinlu3 this is query though. I imagine if user types in `nsclc`, I want it to be matched to NSCLC.

Comment (calvinlu3) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtils.java:117]: Sorry, should clarify that I meant the code from `currSubtype.getKey()`. I assume any future cancer types that we may add to database should not have a lowercase in the code right?

Comment (zhx828) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtils.java:117]: Presumably

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/AnnotationsApiController.java [hunk 4]
```diff
@@ -761,170 +612,9 @@ private IndicatorQueryResp getIndicatorQueryFromHGVSg(
...
-                        if (result == 0) {
-                            // Compare alteration name
-                            if (i1 == null || StringUtils.isNotEmpty(i1.getQuery().getAlteration())) {
-                                return 1;
-                            }
-                            if (i2 == null || StringUtils.isNotEmpty(i2.getQuery().getAlteration())) {
-                                return -1;
-                            }
-                            name1 = i1.getQuery().getAlteration().toLowerCase();
-                            name2 = i2.getQuery().getAlteration().toLowerCase();
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/AnnotationsApiController.java:869]: @calvinlu3 I removed these two lines to use from the parameters. Other changes are directly moved to Utils.

File: core/src/test/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtilsTest.java [hunk 5]
```diff
@@ -47,4 +49,42 @@ public void testSearchNonHgvsAnnotation() {
...
+    public void testAnnotationSearch() {
+        // test gene query
+        Set<AnnotationSearchResult> respSet = annotationSearch("BRAF");
+        // This is a gene search, there should be a gene search result
+        assertTrue(respSet.stream().filter(resp -> AnnotationSearchQueryType.GENE.equals(resp.getQueryType())).count() > 0);
+        // Since BRAF is mentioned in multiple alterations, variant search result should be available
+        assertTrue(respSet.stream().filter(resp -> AnnotationSearchQueryType.VARIANT.equals(resp.getQueryType())).count() > 0);
+        // BRAF is not part of any cancer type, the cancer type search result should not be available
+        assertTrue(respSet.stream().filter(resp -> AnnotationSearchQueryType.CANCER_TYPE.equals(resp.getQueryType())).count() == 0);
+
```
Comment (calvinlu3) [inline core/src/test/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtilsTest.java:62]: Can we add 1 test here for `BCR-ABL1`. There should be gene and alteration results.

Comment (calvinlu3) [reply core/src/test/java/org/mskcc/cbio/oncokb/util/AnnotationSearchUtilsTest.java:62]: The BCR-ABL1 should also have cancer type matches too

**Pull request merged**